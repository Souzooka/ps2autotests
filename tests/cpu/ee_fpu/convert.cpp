#include "shared.h"

#define FF_OP_FUNC(OP, IDENT) \
static inline void FF_##IDENT(register float &fd, const register float &fs) { \
	asm volatile ( \
		OP " %0, %1\n" \
		: "+f"(fd) : "f"(fs) \
	); \
}

static inline void PRINT_ACC(bool newline) {
	static FloatBits result = {0};
	static const float negz = -0.0f;
	static const float posz = 0.0f;

	asm volatile (
		// To grab ACC, we use madd.s 0, 0 -> fd = ACC + 0 * 0.
		// We use negative zero to avoid changing the sign of ACC (note that it's a multiply, -0 * -0 = +0.)
		"madd.s $f4, %0, %1\n"
		"swc1 $f4, 0(%2)\n"
		: : "f"(negz), "f"(posz), "r"((u32)&result.f) : "$f4"
	);

	printf("%08x/", result.u);
	printFloatString(result);
	if (newline) {
		printf("\n");
	}
}

#define FF_OP_DO_II(OP, IDENT, d, s) \
	SET_U32<d>(fd); SET_U32<s>(fs); \
	FF_##IDENT(fd, fs); \
	printf("  %s ", OP); \
	PRINT_F(fs, false); \
	printf(": "); \
	PRINT_F(fd, true);

#define FF_OP_DO_MM(OP, IDENT, d, s) \
	SET_M(fd, d); SET_M(fs, s); \
	FF_##IDENT(fd, fs); \
	printf("  %s %s: ", OP, #s); PRINT_F(fd, true);

#define TEST_FF_FUNC(OP, IDENT) \
FF_OP_FUNC(OP, IDENT); \
static void test_##IDENT() { \
	register float fd, fs; \
	 \
	printf("%s:\n", OP); \
	 \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x00000000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x80000000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x3F800000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x40000000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x40400000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x7FFFFFFF); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xFFFFFFFF); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x7F800000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xFF800000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x00000005); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x0000FFFF); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x4F000000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xCF000000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x4EFFFFFF); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xCEFFFFFF); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x3FBEB852); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x3FC00000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0x3FC147AE); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xBFBEB852); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xBFC00000); \
	FF_OP_DO_II(OP, IDENT, 0x1337, 0xBFC147AE); \
	 \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_ZERO); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_NEGZERO); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_ONE); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_NEGONE); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_MAX_MANTISSA); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_MAX_EXP); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_MIN_EXP); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_MAX); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_MIN); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_GARBAGE1); \
	FF_OP_DO_MM(OP, IDENT, CF_GARBAGE1, CF_GARBAGE2); \
	\
	printf("\n"); \
}

TEST_FF_FUNC("cvt.w.s", cvtws)
TEST_FF_FUNC("cvt.s.w", cvtsw)

int main(int argc, char **argv) {
	printf("-- TEST BEGIN\n");

	test_cvtws();
	test_cvtsw();

	printf("-- TEST END\n");

	return 0;
}
