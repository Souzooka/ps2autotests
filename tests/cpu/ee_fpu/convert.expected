-- TEST BEGIN
cvt.w.s:
  cvt.w.s 00000000/+0.00: 00000000/+0.00
  cvt.w.s 80000000/-0.00: 00000000/+0.00
  cvt.w.s 3f800000/+1.00: 00000001/+0.00
  cvt.w.s 40000000/+2.00: 00000002/+0.00
  cvt.w.s 40400000/+3.00: 00000003/+0.00
  cvt.w.s 7fffffff/+NaN: 7fffffff/+NaN
  cvt.w.s ffffffff/-NaN: 80000000/-0.00
  cvt.w.s 7f800000/+NaN: 7fffffff/+NaN
  cvt.w.s ff800000/-NaN: 80000000/-0.00
  cvt.w.s 00000005/+0.00: 00000000/+0.00
  cvt.w.s 0000ffff/+0.00: 00000000/+0.00
  cvt.w.s 4f000000/+2147483648.00: 7fffffff/+NaN
  cvt.w.s cf000000/-2147483648.00: 80000000/-0.00
  cvt.w.s 4effffff/+2147483520.00: 7fffff80/+NaN
  cvt.w.s ceffffff/-2147483520.00: 80000080/-0.00
  cvt.w.s CF_ZERO: 00000000/+0.00
  cvt.w.s CF_NEGZERO: 00000000/+0.00
  cvt.w.s CF_ONE: 00000001/+0.00
  cvt.w.s CF_NEGONE: ffffffff/-NaN
  cvt.w.s CF_MAX_MANTISSA: 00000001/+0.00
  cvt.w.s CF_MAX_EXP: 7fffffff/+NaN
  cvt.w.s CF_MIN_EXP: 00000000/+0.00
  cvt.w.s CF_MAX: 7fffffff/+NaN
  cvt.w.s CF_MIN: 80000000/-0.00
  cvt.w.s CF_GARBAGE1: 00000000/+0.00
  cvt.w.s CF_GARBAGE2: 80000000/-0.00

cvt.s.w:
  cvt.s.w 00000000/+0.00: 00000000/+0.00
  cvt.s.w 80000000/-0.00: cf000000/-2147483648.00
  cvt.s.w 3f800000/+1.00: 4e7e0000/+1065353216.00
  cvt.s.w 40000000/+2.00: 4e800000/+1073741824.00
  cvt.s.w 40400000/+3.00: 4e808000/+1077936128.00
  cvt.s.w 7fffffff/+NaN: 4effffff/+2147483520.00
  cvt.s.w ffffffff/-NaN: bf800000/-1.00
  cvt.s.w 7f800000/+NaN: 4eff0000/+2139095040.00
  cvt.s.w ff800000/-NaN: cb000000/-8388608.00
  cvt.s.w 00000005/+0.00: 40a00000/+5.00
  cvt.s.w 0000ffff/+0.00: 477fff00/+65535.00
  cvt.s.w 4f000000/+2147483648.00: 4e9e0000/+1325400064.00
  cvt.s.w cf000000/-2147483648.00: ce440000/-822083584.00
  cvt.s.w 4effffff/+2147483520.00: 4e9dffff/+1325399936.00
  cvt.s.w ceffffff/-2147483520.00: ce440000/-822083584.00
  cvt.s.w CF_ZERO: 00000000/+0.00
  cvt.s.w CF_NEGZERO: cf000000/-2147483648.00
  cvt.s.w CF_ONE: 4e7e0000/+1065353216.00
  cvt.s.w CF_NEGONE: ce810000/-1082130432.00
  cvt.s.w CF_MAX_MANTISSA: 4e7fffff/+1073741760.00
  cvt.s.w CF_MAX_EXP: 4eff0000/+2139095040.00
  cvt.s.w CF_MIN_EXP: 3f800000/+1.00
  cvt.s.w CF_MAX: 4effffff/+2147483520.00
  cvt.s.w CF_MIN: bf800000/-1.00
  cvt.s.w CF_GARBAGE1: 4599b800/+4919.00
  cvt.s.w CF_GARBAGE2: ce054904/-559038720.00

-- TEST END
